### PR TITLE
expr/builtin_fncall: implement Inet6Ntoa

### PR DESCRIPTION
**Pull request by sweetIan**
Signed-off-by: sweetIan <ian.d2d@qq.com>

## What have you changed? (mandatory)

expr/builtin_fncall: implement Inet6Ntoa

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 1]
```diff
@@ -39,6 +39,35 @@ impl ScalarFunc {
...
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        if s.len() == 16 {
+            let mut v = [0 as u16; 8];
+            for i in 0..8 {
+                v[i] = u16::from(s[i * 2]) << 8 | u16::from(s[i * 2 + 1]);
+            }
+            Ok(Some(Cow::Owned(
+                format!(
+                    "{}",
+                    Ipv6Addr::new(v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])
```
[the remark this answers is not in the document]
Comment (sweetIan) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: IpAddr::from_str converts "x.x.x.x", "xxxx::xxxx.xxxx",etc. to IpAdrr. But here we are doing the reverse. I don't understand why we should use IpAddr::from_str.

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 2]
```diff
@@ -39,6 +39,35 @@ impl ScalarFunc {
...
     }
+
+    pub fn inet6_ntoa<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        if s.len() == 16 {
+            let mut v = [0 as u16; 8];
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: `as u16` is redundant.

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 3]
```diff
@@ -39,6 +39,35 @@ impl ScalarFunc {
...
+            for i in 0..8 {
+                v[i] = u16::from(s[i * 2]) << 8 | u16::from(s[i * 2 + 1]);
+            }
+            Ok(Some(Cow::Owned(
+                format!(
+                    "{}",
+                    Ipv6Addr::new(v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])
+                ).as_bytes()
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: `into_bytes` seems better.

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: How about
```
let v: &[u8; 16] = s.as_ref().try_into().unwrap();
Ipv6Addr::from(v.clone())
```
?

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 4]
```diff
@@ -39,6 +40,27 @@ impl ScalarFunc {
...
+    pub fn inet6_ntoa<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        if s.len() == 16 {
+            let v: &[u8; 16] = s.as_ref().try_into().unwrap();
+            Ok(Some(Cow::Owned(
+                format!("{}", Ipv6Addr::from(*v)).into_bytes().to_vec(),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: `to_vec` is not necessary.

**Comment by huachaohuang** [comment]
@solotzg @BusyJay PTAL

**Pull request merged**